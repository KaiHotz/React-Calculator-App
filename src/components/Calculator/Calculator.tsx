import React from 'react';
import _ from 'lodash';
import { CalculatorDisplay } from '../CalculatorDisplay';
import { CalculatorKey } from '../CalculatorKey';
import { calculatorOperations, digitKeys } from '../../utils/helpers';
import { useCalculator } from '../../hooks/useCalculator';
import { EInputTypes } from '../../types';
import './Calculator.scss';

export const Calculator = () => {
  const { state, handleClick } = useCalculator();

  return (
    <div className="calculator">
      <CalculatorDisplay value={state.displayValue} />
      <div className="calculator-keypad">
        <div className="input-keys">
          <div className="function-keys">
            <CalculatorKey
              className="key-clear"
              onClick={() => handleClick(state.displayValue !== '0' ? EInputTypes.clearDisplay : EInputTypes.clearAll)}
              keyValue={state.displayValue !== '0' ? 'C' : 'AC'}
            />
            <CalculatorKey className="key-sign" onClick={() => handleClick(EInputTypes.toggleSign)} keyValue="±" />
            <CalculatorKey className="key-percent" onClick={() => handleClick(EInputTypes.inputPercent)} keyValue="%" />
          </div>
          <div className="digit-keys">
            {digitKeys.map((digit) => (
              <CalculatorKey
                key={`key-${digit}`}
                className={`key-${digit}`}
                onClick={() => handleClick(EInputTypes.inputDigit, digit)}
                keyValue={digit}
              />
            ))}
            <CalculatorKey className="key-dot" onClick={() => handleClick(EInputTypes.inputDot)} keyValue="●" />
          </div>
        </div>
        <div className="operator-keys">
          {_.map(calculatorOperations, (value, key) => (
            <CalculatorKey
              key={`key-${value.name}`}
              className={`key-${value.name}`}
              onClick={() => handleClick(EInputTypes.performOperation, key)}
              keyValue={value.symbol}
            />
          ))}
        </div>
      </div>
    </div>
  );
};
