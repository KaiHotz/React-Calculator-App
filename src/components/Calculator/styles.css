.calculator {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  box-shadow: 0px 0px 20px 0px #aaa;
}

.calculator .input-keys {
  width: 240px;
}

.calculator .function-keys {
  display: flex;
}

.calculator .digit-keys {
  background: #e0e0e7;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
}

.calculator-keypad {
  height: 350px;
  display: flex;
}

.calculator .function-keys .calculator-key {
  font-size: 2em;
}
.calculator .function-keys .key-multiply {
  line-height: 50px;
}
.calculator .digit-keys .calculator-key {
  font-size: 2.25em;
}
.calculator .digit-keys .key-0 {
  width: 160px;
  text-align: left;
  padding-left: 32px;
}
.calculator .digit-keys .key-dot {
  padding-top: 1em;
  font-size: 0.75em;
}
.calculator .operator-keys .calculator-key {
  color: white;
  border-right: 0;
  font-size: 3em;
}

.calculator .function-keys {
  background:  rgba(202,202,204,1);
}
.calculator .operator-keys {
  background:  rgba(252,156,23,1);
}

