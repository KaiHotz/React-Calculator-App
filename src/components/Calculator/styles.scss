.calculator {
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: column;
  box-shadow: 0px 0px 20px 0px #aaa;
  .calculator-keypad {
    height: 350px;
    display: flex;
  }
  .input-keys {
    width: 240px;
  }
  .function-keys {
    display: flex;
    background: linear-gradient(to bottom, rgba(202,202,204,1) 0%, rgba(196,194,204,1) 100%);
    .calculator-key {
      font-size: 2em;
    }
    .key-multiply {
      line-height: 50px;
    }
  }
  .digit-keys {
    background: #e0e0e7;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    .calculator-key {
      font-size: 2.25em;
    }
    .key-0 {
      width: 160px;
      text-align: left;
      padding-left: 32px;
    }
    .key-dot {
      padding-top: 1em;
      font-size: 0.75em;
    }
  }
  .operator-keys{
    background:  linear-gradient(to bottom, rgba(252,156,23,1) 0%, rgba(247,126,27,1) 100%);
    .calculator-key {
      color: white;
      border-right: 0;
      font-size: 3em;
    }
  }

}

