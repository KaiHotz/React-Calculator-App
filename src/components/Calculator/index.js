import React from 'react'
import _ from 'lodash'
import CalculatorDisplay from '../CalculatorDisplay'
import CalculatorKey from '../CalculatorKey'
import { CalculatorOperations, DigitKeys } from '../../utils/helper'
import { useCalculator } from '../../hooks'
import './styles.scss'

const Calculator = () => {
  const [state, handleClick] = useCalculator()

  return (
    <div className="calculator">
      <CalculatorDisplay value={state.displayValue} />
      <div className="calculator-keypad">
        <div className="input-keys">
          <div className="function-keys">
            <CalculatorKey
              className="key-clear"
              onClick={handleClick(state.displayValue !== '0' ? 'clearDisplay' : 'clearAll')}
            >
              {state.displayValue !== '0' ? 'C' : 'AC'}
            </CalculatorKey>
            <CalculatorKey
              className="key-sign"
              onClick={handleClick('toggleSign')}
            >
              ±
            </CalculatorKey>
            <CalculatorKey
              className="key-percent"
              onClick={handleClick('inputPercent')}
            >
              %
            </CalculatorKey>
          </div>
          <div className="digit-keys">
            {
              DigitKeys.map(digit => (
                <CalculatorKey
                  key={`key-${digit}`}
                  className={`key-${digit}`}
                  onClick={handleClick('inputDigit', digit)}
                >
                  {digit}
                </CalculatorKey>
              ))
            }
            <CalculatorKey
              className="key-dot"
              onClick={handleClick('inputDot')}
            >
              ●
            </CalculatorKey>
          </div>
        </div>
        <div className="operator-keys">
          {
            _.map(CalculatorOperations, (value, key) => (
              <CalculatorKey
                key={`key-${value.name}`}
                className={`key-${value.name}`}
                onClick={handleClick('performOperation', key)}
              >
                {value.symbol}
              </CalculatorKey>
            ))
          }
        </div>
      </div>
    </div>
  )
}

export default Calculator
