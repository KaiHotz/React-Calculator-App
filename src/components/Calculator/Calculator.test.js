import React from 'react'
import { shallow } from 'enzyme'

import Calculator from './index'

describe('<Calculator />', () => {
  it('should render', () => {
    const wrapper = shallow(<Calculator />)

    expect(wrapper).toBeDefined()
  })

  /* Tests will be adjusted upon Enzyme update to support Hooks */

  // it('should sum correctly', () => {
  //   const wrapper = mount(<Calculator />)
  //   wrapper.find('.calculator-key.key-6').simulate('click')
  //   wrapper.find('.calculator-key.key-add').simulate('click')
  //   wrapper.find('.calculator-key.key-2').simulate('click')
  //   wrapper.find('.calculator-key.key-equals').simulate('click')

  //   expect(wrapper.state().displayValue).toEqual('8')
  // })

  // it('should rest correctly', () => {
  //   const wrapper = mount(<Calculator />)
  //   wrapper.find('.calculator-key.key-6').simulate('click')
  //   wrapper.find('.calculator-key.key-subtract').simulate('click')
  //   wrapper.find('.calculator-key.key-2').simulate('click')
  //   wrapper.find('.calculator-key.key-equals').simulate('click')

  //   expect(wrapper.state().displayValue).toEqual('4')
  // })

  // it('should divide correctly', () => {
  //   const wrapper = mount(<Calculator />)
  //   wrapper.find('.calculator-key.key-6').simulate('click')
  //   wrapper.find('.calculator-key.key-divide').simulate('click')
  //   wrapper.find('.calculator-key.key-2').simulate('click')
  //   wrapper.find('.calculator-key.key-equals').simulate('click')

  //   expect(wrapper.state().displayValue).toEqual('3')
  // })

  // it('should multiply correctly', () => {
  //   const wrapper = mount(<Calculator />)
  //   wrapper.find('.calculator-key.key-6').simulate('click')
  //   wrapper.find('.calculator-key.key-multiply').simulate('click')
  //   wrapper.find('.calculator-key.key-2').simulate('click')
  //   wrapper.find('.calculator-key.key-equals').simulate('click')

  //   expect(wrapper.state().displayValue).toEqual('12')
  // })

  // it('should show decimals correctly', () => {
  //   const wrapper = mount(<Calculator />)
  //   wrapper.find('.calculator-key.key-dot').simulate('click')
  //   wrapper.find('.calculator-key.key-2').simulate('click')

  //   expect(wrapper.state().displayValue).toEqual('0.2')
  // })

  // it('should invert sign correctly', () => {
  //   const wrapper = mount(<Calculator />)
  //   wrapper.find('.calculator-key.key-2').simulate('click')
  //   wrapper.find('.calculator-key.key-sign').simulate('click')

  //   expect(wrapper.state().displayValue).toEqual('-2')
  // })

  // it('should apply % correctly', () => {
  //   const wrapper = mount(<Calculator />)
  //   wrapper.find('.calculator-key.key-2').simulate('click')
  //   wrapper.find('.calculator-key.key-percent').simulate('click')

  //   expect(wrapper.state().displayValue).toEqual('0.02')
  // })

  // it('should clear the display correctly', () => {
  //   const wrapper = mount(<Calculator />)
  //   wrapper.find('.calculator-key.key-2').simulate('click')
  //   wrapper.find('.calculator-key.key-3').simulate('click')
  //   wrapper.find('.calculator-key.key-4').simulate('click')
  //   wrapper.find('.calculator-key.key-clear').simulate('click')

  //   expect(wrapper.state().displayValue).toEqual('0')
  // })

  // it('should toggle the sign correctly', () => {
  //   const wrapper = mount(<Calculator />)
  //   wrapper.find('.calculator-key.key-2').simulate('click')
  //   wrapper.find('.calculator-key.key-sign').simulate('click')

  //   expect(wrapper.state().displayValue).toEqual('-2')
  // })

  // it('should clear the display', () => {
  //   const wrapper = mount(<Calculator />)
  //   wrapper.find('.calculator-key.key-2').simulate('click')
  //   wrapper.find('.calculator-key.key-multiply').simulate('click')
  //   wrapper.find('.calculator-key.key-3').simulate('click')
  //   wrapper.find('.calculator-key.key-clear').simulate('click')

  //   expect(wrapper.state()).toEqual({
  //     value: 2,
  //     displayValue: '0',
  //     operator: '*',
  //     waitingForOperand: false,
  //   })
  // })

  // it('should clear all', () => {
  //   const wrapper = mount(<Calculator />)
  //   wrapper.find('.calculator-key.key-2').simulate('click')
  //   wrapper.find('.calculator-key.key-multiply').simulate('click')
  //   wrapper.find('.calculator-key.key-3').simulate('click')
  //   wrapper.find('.calculator-key.key-equals').simulate('click')
  //   wrapper.find('.calculator-key.key-clear').simulate('click')
  //   wrapper.find('.calculator-key.key-clear').simulate('click')

  //   expect(wrapper.state()).toEqual({
  //     value: null,
  //     displayValue: '0',
  //     operator: null,
  //     waitingForOperand: false,
  //   })
  // })
})
