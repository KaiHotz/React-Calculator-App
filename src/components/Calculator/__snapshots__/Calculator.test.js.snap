// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Calculator /> should render 1`] = `ShallowWrapper {}`;
